Intro to api design (#6)
Ada-Developers-Academy/core-unit-2: intro-to-api-design/designing-endpoints.md

# Designing Endpoints

## Learning Goals

- Define REST as an architecture style for applications/APIs that uses the HTTP protocol and uses client-server architecture
- Define the REST principle of "uniform interface"
- List all HTTP Request methods: GET, HEAD, POST, PUT, DELETE, TRACE, OPTIONS, CONNECT, PATCH
- Define CRUD as a pattern of four basic functions of persisted data
- Compare the REST principle of "uniform interface" with CRUD and the HTTP Request methods

## Introduction

Imagine Eryn, a developer who's making the back-end API for an educational podcast app. The app will allow users to sign up as students or instructors. Students can browse different courses. Each course will have several class sessions, which will contain podcast episodes.

> Eryn's current challenge is making the API that will give back the data about the episodes, class sessions, courses, instructors, and students. Eryn needs to create endpoints, or entry ways, so clients can send requests to her API and get back the data they need. But what makes a good endpoint, or good entryway into her API? How does Eryn know that she's making the right endpoints?

Reviewer: ```suggestion
Eryn's current challenge is making the API that will give back the data about the episodes, class sessions, courses, instructors, and students. Eryn needs to create endpoints, or entryways, so clients can send requests to her API and get back the data they need. But what makes a good endpoint, or a good entryway into her API? How does Eryn know that she's making the right endpoints?
```

| -------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------- |
| Resource-based Paths                         | Request URLs and paths should be based around a _resource_. Recall that resources are any piece of data to return, such as user data or web pages. | If there is a resource about dog data, `/dogs` |
| Manipulate Resources Through Representations | Clients can modify or delete resources on a server if they have a representation of it, such as an ID                                              | Deleting the Dog resource of ID 45             |
| Self-Descriptive Messages                    | HTTP requests and responses should include the information of how to be read, by carrying information about their format (usually as a header)     |

Let's revisit REST's name: "Representational state transfer." After considering the name and these guidelines, we may conclude:

> - REST prioritizes _resources_ and _representations of resources_, and this will effect how URLs are made, what information to send with HTTP requests, and what we can expect from HTTP responses.

Reviewer: ```suggestion
- REST prioritizes _resources_ and _representations of resources_, and this will affect how URLs are made, what information to send with HTTP requests, and what we can expect from HTTP responses.
```

- REST prioritizes _statelessness_ on the server between requests. It maintains statelessness between requests by relying on HTTP requests to send any details about client state.

## Meaningful Endpoints to Access Resources

> Recall that an _endpoint_ is an entrypoint URL to access a web API. For example, in the LocationIQ API, `/search` is the endpoint to access the API's search functionality.

Reviewer: ```suggestion
Recall that an _endpoint_ is an entry point URL to access a web API. For example, in the LocationIQ API, `/search` is the endpoint to access the API's search functionality.
```

Also recall that resources are any piece of data to return, such as user data or web pages.

A well-designed RESTful API states that endpoints should be based on resources.

> Imagine Eryn, who's making the back-end API for an educational podcast app. Eryn's educational app stores data about every student user, every instructor user, every different courses offered, and all class sessions within the courses.

Reviewer: ```suggestion
Imagine Eryn, who's making the back-end API for an educational podcast app. Eryn's educational app stores data about every student user, every instructor user, every different course offered, and all class sessions within the courses.
```

Some example endpoints for Eryn may be:

- `/students`
- `/instructors`
- `/courses`
- `/classes`

## Meaningful HTTP Request Methods

> The REST principle "self-descriptive messages" includes the princple of meaningful HTTP request methods. Well-designed RESTful APIs use HTTP request method in combination with resource-centered endpoints.

Reviewer: ```suggestion
The REST principle "self-descriptive messages" includes the principle of meaningful HTTP request methods. Well-designed RESTful APIs use the HTTP request method in combination with resource-centered endpoints.
```
